dbms/src/Functions/FunctionsGrouping.h

```diff
@@ -146,31 +158,58 @@ class FunctionGrouping : public IFunctionBase
 
     ResultType groupingImplModeAndBit(UInt64 grouping_id) const
     {
-        return (grouping_id & meta_grouping_id) != 0;
+        UInt64 final = 0;
+        for (auto one_grouping_id : meta_grouping_ids)
+        {
+            final <<= 1;
+            if ((grouping_id & one_grouping_id) > 0)
+                // col is not needed, meaning being filled null and grouped. = 1
+                final += 1;
+        }
+        return final;
     }
 
     ResultType groupingImplModeNumericCmp(UInt64 grouping_id) const
     {
-        return grouping_id > meta_grouping_id;
+        UInt64 final = 0;
+        for (auto one_grouping_id : meta_grouping_ids)
+        {
+            final <<= 1;
+            if (grouping_id > one_grouping_id)
+                // col is not needed, meaning being filled null and grouped. = 1
+                final += 1;
+        }
+        return final;
     }
 
     ResultType groupingImplModeNumericSet(UInt64 grouping_id) const
     {
-        auto iter = meta_grouping_marks.find(grouping_id);
-        return iter == meta_grouping_marks.end();
+        UInt64 final = 0;
+        for (auto one_grouping_mark : meta_grouping_marks)
+        {
+            final <<= 1;
+            auto iter = one_grouping_mark.find(grouping_id);
+            if (iter != one_grouping_mark.end())
+                // In num-set mode, grouping marks stores those not-needed-col's grouping set (GIDs).
+                // When we can find the grouping id in set, it means this col is not needed and has been filled with null and grouped. = 1
+                final += 1;
+        }
+        return final;
     }
 
 private:
     DataTypes argument_types;
     DataTypePtr return_type;
 
     tipb::GroupingMode mode;
-    UInt64 meta_grouping_id = 0;
+    // one more dimension for multi grouping function args like: grouping(x,y,z...)
+    std::vector<UInt64> meta_grouping_ids;
 
     // In grouping function, the number of rolled up columns usually very small,
     // so it's appropriate to use std::set as it is faster than unordered_set in
     // small amount of elements.
-    std::set<UInt64> meta_grouping_marks = {};
+    // one more dimension for multi grouping function args like: grouping(x,y,z...)
+    std::vector<std::set<UInt64>> meta_grouping_marks = {};
 };
 
 class FunctionBuilderGrouping : public IFunctionBuilder
```

Comment:
In `getReturnTypeImpl` function, argument number should be re-considered.

Reply:
as said in the comment above, the pb function that tiflash received only has one arg (gid) with metas, so `getReturnTypeImpl` seems good